### PR TITLE
Add webhooks test

### PR DESCRIPTION
**Pull request by mkudlej**
This PR adds testcase for testing k8s Dynamic Admission Control webhooks, see https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/ which are used for calling Authorino service for authenticating of Ingress object creation.

### CODE REVIEW

File: Pipfile [hunk 1]
```diff
@@ -4,9 +4,9 @@ pylint = "*"
 types-PyYAML = "*"
 # Have commented out python-language-server to make it available quickly
 # for the development
-#python-language-server = "*"
-#pyls-mypy = "*"
-#pyls-isort = "*"
+# python-language-server = "*"
+# pyls-mypy = "*"
+# pyls-isort = "*"
```
Comment (JaurbanRH) [inline Pipfile:9]: Unrelated change

File: testsuite/objects/__init__.py [hunk 2]
```diff
@@ -67,7 +67,7 @@ def add_oidc_identity(self, name, endpoint, credentials, selector):
         """Adds OIDC identity provider"""
 
     @abc.abstractmethod
-    def add_api_key_identity(self, name, all_namespaces, match_label, match_expression,  credentials, selector):
+    def add_api_key_identity(self, name, all_namespaces, match_label, match_expression, credentials, selector):
```
Comment (JaurbanRH) [inline testsuite/objects/__init__.py:70]: Unrelated change

File: testsuite/openshift/objects/authorino.py [hunk 3]
```diff
@@ -50,6 +51,8 @@ def create_instance(cls, openshift: OpenShiftClient, name, image=None,
                 "enabled": True,
                 "certSecretRef": {"name": listener_certificate_secret}
             }
+        if specific_name:
+            model["metadata"]["name"] = specific_name
```
Comment (JaurbanRH) [inline testsuite/openshift/objects/authorino.py:55]: You can use `name` attribute instead of creating a new one that is for same thing.

File: testsuite/tests/kuadrant/authorino/operator/tls/test_webhook.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        })
+    }
+    return cert_builder(cfssl, chain, wildcard_domain)
+
+
+@pytest.fixture(scope="module")
+def authorino_parameters(authorino_parameters, specific_authorino_name):
+    """Setup TLS with specific name for authorino."""
+    authorino_parameters['specific_name'] = specific_authorino_name
+    yield authorino_parameters
```
Comment (JaurbanRH) [inline testsuite/tests/kuadrant/authorino/operator/tls/test_webhook.py:52]: Why `yield` instead of `return` ?

File: testsuite/openshift/objects/authorino.py (hunk 3 above)
Comment (mkudlej) [reply testsuite/openshift/objects/authorino.py:55]: For 99% of testcases is OK to have Authorino with blamed name as it is generated here https://github.com/Kuadrant/testsuite/blob/main/testsuite/tests/kuadrant/authorino/conftest.py#L34
Only this testcase needs to have it. But you are right that this is not good solution, so I've found better place to set the Authorino name here testsuite/tests/kuadrant/authorino/conftest.py

Comment (pehala) [reply testsuite/openshift/objects/authorino.py:55]: You shouldnt change it in upper conftest either change it properl here and it cannot be done then we need to create a way how to do it here

File: testsuite/tests/kuadrant/authorino/operator/tls/test_webhook.py (hunk 4 above)
Comment (mkudlej) [reply testsuite/tests/kuadrant/authorino/operator/tls/test_webhook.py:52]: `yield` was there before, see https://github.com/Kuadrant/testsuite/blob/main/testsuite/tests/kuadrant/authorino/operator/tls/conftest.py#L160 but I've changed both.

File: testsuite/openshift/objects/validating_webhook.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    def create_instance(cls, openshift: 'OpenShiftClient', name: str, service_name: str, ca_bundle: str):
+        """Creates base instance"""
+        model: Dict[str, Any] = {
+            'apiVersion': 'admissionregistration.k8s.io/v1',
+            'kind': 'ValidatingWebhookConfiguration',
+            'metadata': {
+                'name': name,
+                'namespace': openshift.project
+            },
+            'webhooks': [
```
[the remark this answers is not in the document]
Comment (mkudlej) [reply testsuite/openshift/objects/validating_webhook.py:25]: it will not change I think

[the remark this answers is not in the document]
Comment (pehala) [reply testsuite/openshift/objects/validating_webhook.py:25]: If it doesnt need to change and is only useful for one test it doesn't need a class

File: testsuite/tests/kuadrant/authorino/operator/tls/test_webhook.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+    """
+    Hostname of the upstream certificate sent to be validated by APIcast
+    May be overwritten to configure different test cases
+    """
+    return f"{specific_authorino_name}-authorino-authorization.{openshift.project}.svc"
+
+
+@pytest.fixture(scope="session")
+def certificates(cfssl, authorino_domain, wildcard_domain):
+    """Certificate hierarchy used for the tests"""
```
[the remark this answers is not in the document]
Comment (mkudlej) [reply testsuite/tests/kuadrant/authorino/operator/tls/test_webhook.py:32]: The big difference is that Authorino certificate has `hosts` value to value of `authorino_domain` fixture.

File: testsuite/tests/kuadrant/authorino/operator/tls/test_webhook.py [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+        ingress := request.oldObject { verb == "DELETE" }
+        forbidden { count(object.get(ingress.spec, "rules", [])) == 0 }
+        rules { count(ingress.spec.rules) == 1; ingress.spec.rules[0] == {} }
+        allow { rules; not forbidden }
+    """
+    authorization.add_opa_policy('features', opa_policy)
+
+    when = [{'selector': 'auth.authorization.features.allow', 'operator': 'eq', 'value': 'true'},
+            {'selector': 'auth.authorization.features.verb', 'operator': 'eq', 'value': 'CREATE'}]
+    kubernetes = {
```
[the remark this answers is not in the document]
Comment (mkudlej) [reply testsuite/tests/kuadrant/authorino/operator/tls/test_webhook.py:79]: I don't have the same feeling :) There are just 2 short dicts. When there is a need to have complex `when` condition we can create wrapper for it with special treatment.

File: testsuite/tests/kuadrant/authorino/operator/tls/test_webhook.py [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+# pylint: disable=unused-argument
+@pytest.fixture(scope="module")
+def authorization(authorization, openshift, module_label, authorino_domain) -> Authorization:
+    """In case of Authorino, AuthConfig used for authorization"""
+    authorization.remove_all_hosts()
+    authorization.add_host(authorino_domain)
+
+    authorization.remove_all_identities()
+    plain = {'authJSON': 'context.request.http.body.@fromstr|request.userInfo'}
+    authorization.add_kubernetes_identity('k8s-userinfo', plain)
```
[the remark this answers is not in the document]
Comment (mkudlej) [reply testsuite/tests/kuadrant/authorino/operator/tls/test_webhook.py:64]: Maybe I don't know what you exactly mean. Validation webhook request contains various information which ca be used but for this test this one is the only one used. Should I move it to separated function then?

[the remark this answers is not in the document]
Comment (pehala) [reply testsuite/tests/kuadrant/authorino/operator/tls/test_webhook.py:64]: `authorization.add_kubernetes_identity('k8s-userinfo', 'context.request.http.body.@fromstr|request.userInfo')` are the input parameters I would like to see (you can use variables ofc)

File: testsuite/tests/kuadrant/authorino/conftest.py [hunk 9]
```diff
@@ -30,8 +30,10 @@ def authorino(authorino, openshift, blame, request, testconfig, module_label, au
     labels = authorino_parameters.setdefault("label_selectors", [])
     labels.append(f"testRun={module_label}")
 
+    authorino_name = authorino_parameters.pop('specific_name', blame("authorino"))
```
Comment (pehala) [inline testsuite/tests/kuadrant/authorino/conftest.py:33]: `authorino_parameters.setdefault("name", blame("authorino"))`

File: testsuite/openshift/objects/authorino.py [hunk 10]
```diff
@@ -14,8 +14,8 @@ class AuthorinoCR(OpenShiftObject, Authorino):
 
     @classmethod
     def create_instance(cls, openshift: OpenShiftClient, name, image=None,
-                        cluster_wide=False, label_selectors: List[str] = None, listener_certificate_secret=None,
-                        log_level=None):
```
Comment (pehala) [inline testsuite/openshift/objects/authorino.py:18]: Unrelated change

File: testsuite/tests/kuadrant/authorino/operator/tls/test_webhook.py [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+
+# pylint: disable=unused-argument
+@pytest.fixture(scope="module")
+def authorization(authorization, openshift, module_label, authorino_domain) -> Authorization:
+    """In case of Authorino, AuthConfig used for authorization"""
+    authorization.remove_all_hosts()
+    authorization.add_host(authorino_domain)
+
+    authorization.identity.remove_all()
+    plain = {'authJSON': 'context.request.http.body.@fromstr|request.userInfo'}
```
Comment (pehala) [inline testsuite/tests/kuadrant/authorino/operator/tls/test_webhook.py:74]: I would change the interface to just support `plain = 'context.request.http.body.@fromstr|request.userInfo'`

File: testsuite/tests/kuadrant/authorino/operator/tls/test_webhook.py [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+    authorization.identity.kubernetes('k8s-userinfo', plain)
+
+
+    authorization.authorization.opa_policy('features', OPA_POLICY)
+
+    when = [{'selector': 'auth.authorization.features.allow', 'operator': 'eq', 'value': 'true'},
+            {'selector': 'auth.authorization.features.verb', 'operator': 'eq', 'value': 'CREATE'}]
+    kube_user = {'valueFrom': {'authJSON': 'auth.identity.username'}}
```
Comment (pehala) [inline testsuite/tests/kuadrant/authorino/operator/tls/test_webhook.py:82]: Similar here, you were right that there is an option to use constant value but there will most likely be a future refactor that will unify the behaviour of `valueFrom` and `value` and if we cut it down to just `auth.identity.username` I think it will be more readable and easier to work with

File: testsuite/tests/kuadrant/authorino/operator/tls/test_webhook.py [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+                 }
+    authorization.authorization.kubernetes('ingress-authn-k8s-binding-delete',
+                                           when, kube_user, kube_attrs, 1)
+    return authorization
+
+
+@pytest.fixture(scope="module")
+def validating_webhook(openshift, authorino_domain, certificates):
+    """Create validating webhook."""
+    name = 'check-ingress.authorino.kuadrant.io'
```
Comment (pehala) [inline testsuite/tests/kuadrant/authorino/operator/tls/test_webhook.py:109]: What will happen if two people run this test in parallel? Wouldnt it break because the name is not randomised?

File: testsuite/tests/kuadrant/authorino/operator/tls/test_webhook.py [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+                                                certificates['authorino_ca'])
+    webhook.commit()
+    yield webhook
+    webhook.delete()
+
+
+# pylint: disable=unused-argument
+def test_authorized_via_http(authorization, openshift, authorino, authorino_domain, validating_webhook):
+    """Test raw http authentization via webhooks."""
+    ingress = Ingress.create_instance(openshift, 'minimal-ingress', rules=[{}])
```
Comment (pehala) [inline testsuite/tests/kuadrant/authorino/operator/tls/test_webhook.py:121]: Similar problem (parallel runs) here

File: testsuite/openshift/objects/validating_webhook.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (pehala) [reply testsuite/openshift/objects/validating_webhook.py:29]: > If it doesnt need to change and is only useful for one test it doesn't need a class

This is still valid, in current state ValidatingWebhook class is not needed as it is hardcoded to Ingresses only and doesn't add any additional value, this can be a simple fixture and everything would work the same

**Review by pehala — COMMENTED** [review]
Thanks for the changes! It is much better now!

**Review by pehala — DISMISSED** [review]

**Comment by jsmolar** [comment]
@mkudlej PR looks good, but I would appreciate a short description for `test_webhook.py` that briefly summarizes the tests case (mainly setup)

**Comment by mkudlej** [comment]
@jsmolar Is it better?

**Review by pehala — APPROVED** [review]
